Explain how to add Swagger REST API doc coverage in zulip.yaml.

feorlen opened the pull request:
When #3397 is resolved and folks start using the Swagger interface, we can
determine where this new doc should go and if it needs to be more
comprehensive.


Introduce Swagger UI and the Swagger/OpenAPI specification. Explain
the structure of zulip.yaml and show examples of different sections
of the file.

This is a new file in /docs not yet included in the Read the Docs
table of contents. Where it should go should be determined as we iterate
on the Swagger UI integration and expand REST API doc coverage using it.

For more on Swagger UI and how Zulip uses it, see:
https://github.com/zulip/zulip/issues/3474
https://github.com/zulip/zulip/pull/3397

smarx commented:
*Automated message from Dropbox CLA bot*

@feorlen, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on docs/swagger-doc-coverage.md:
```diff
@@ -0,0 +1,178 @@
+defined according to the Swagger/OpenAPI specification. For Zulip, this is in
+`/static/yaml/zulip.yaml`, and the generated API reference is available at a URL
+similar to other Zulip documentation. You can find an example at
+[/apidocs](https://zulipchat.com/apidocs/). This page is a basic introduction to the
```
Maybe instead suggest the location in development, e.g. `http://localhost:9991/apidocs/`?  That's what they'll want to look at in order to iterate on it in development.

timabbott commented on docs/swagger-doc-coverage.md:
````diff
@@ -0,0 +1,178 @@
+          description: Unexpected error
+          schema:
+            $ref: '#/definitions/ErrorModel'
+```
````
I would consider adding adding a first example of something that doesn't require arguments at all, e.g. `GET /users`, just to show what a really simple one looks like.

timabbott commented:
@YagoGG can you give this PR a review, based on your experiences with teaching people Swagger at the PyCon sprints?

@feorlen it'd be awesome if you can resolve the comments I posted above; I'd love to get this doc merged.

YagoGG commented:
I think this is great, very nicely explained! Just a few comments:

- We should probably include a "styling" section, explaining how we like the syntax in the YAML file (e.g. single quotes, no indentation in lines starting with a dash...).
- Explaining how to use "inheritance" in the definitions (with `allOf`) would avoid duplication and definitely save a lot of time while reviewing API docs PRs.
- Since we aren't using Swagger UI (at least for now), we should probably remove that part before merging.

Thanks for writing this, @feorlen!

YagoGG commented on docs/swagger-doc-coverage.md:
```diff
@@ -0,0 +1,178 @@
+a red "ERROR" button at the bottom of the page instead of the green "VALID" one
+even if your file is correct.
+
+Formatting help:
```
Perhaps we want this to be a third level header?

YagoGG commented on docs/swagger-doc-coverage.md:
```diff
@@ -0,0 +1,178 @@
+  Single spaced lines (one newline at the end of each) are joined. Use an extra
+  blank line for a paragraph break.
+
+Examples:
```
Same here.

YagoGG commented on docs/swagger-doc-coverage.md:
````diff
@@ -0,0 +1,178 @@
+are expected as POST data, as you might do with `curl -d`.
+
+```
+paths:
````
Mind updating the example with the latest changes in the YAML file? There are a couple things that have changed, and it could be confusing for beginners :)

feorlen commented:
@YagoGG Thanks for the comments! Some of this has been pending getting the latest Swagger UI version sorted out. Now with the additional stuff you added in the yaml file, there's more that could be said. Swagger UI I think will still be useful, but now this doc should be less tied to it particularly.

feorlen commented:
I reworked several areas to make the document more general. I replaced the /messages example with /users/{user}/presence because it was something already in zulip.yaml that has only GET. Are there other examples from the file that would be helpful to describe?

@YagoGG I made a new empty section for style, so I can start adding local preferences there. I presume we are pretty set on 2 space indents, you mentioned a few others, what other style notes would be useful in addition? 

How soon do you expect to start breaking the one zulip.yaml into multiple files? I could look into the mechanics of that and describe it more if that would be useful.

timabbott commented:
I think leaving the concept of splitting the file for future work is probably the right path.

timabbott commented:
This looks great!  I made a few changes:
* Renamed the file to `swagger-api-docs.md`
* Added a link from the TOC
* Added a link from the `zulip.yaml` file

and merged this.  Thanks @feorlen!  

@YagoGG I encourage you to give this a read and submit a follow-up PR for tweaks on things like expanding the style guide; I'm merging this because I think it's in a reasonable state, not because I think this is the final version we'll have (even in a week or two).
